net/ipv4/tcp_bic_rust.rs

```diff
@@ -0,0 +1,296 @@
+//! SPDX-License-Identifier: GPL-2.0
+//!
+//! Binary Increase Congestion control (BIC). Based on:
+//!     Binary Increase Congestion Control (BIC) for Fast Long-Distance
+//!     Networks - Lisong Xu, Khaled Harfoush, and Injong Rhee
+//!     IEEE INFOCOM 2004, Hong Kong, China, 2004, pp. 2514-2524 vol.4
+//!     doi: 10.1109/INFCOM.2004.1354672
+
+use core::cmp::{max, min};
+use core::num::NonZeroU32;
+use kernel::net::tcp::cong;
+use kernel::prelude::*;
+use kernel::time;
+use kernel::{c_str, module_cca};
+
+const ACK_RATIO_SHIFT: u32 = 4;
+
+// TODO: Convert to module parameters once they are available.
+/// Value of ssthresh for new connections.
+const INITIAL_SSTHRESH: Option<u32> = None;
+/// If cwnd is larger than this threshold, BIC engages; otherwise normal TCP
+/// increase/decrease will be performed.
+// NOTE: cwnd is expressed in units of full-sized segments.
+const LOW_WINDOW: u32 = 14;
+/// In binary search, go to point: `cwnd + (W_max - cwnd) / BICTCP_B`.
+// SAFETY: Obvious.
+// TODO: Convert to `new::(x).unwrap()` once 'const_option' is stabilised.
+const BICTCP_B: NonZeroU32 = unsafe { NonZeroU32::new_unchecked(4) };
```

Comment:
Note that you will need a `// SAFETY: ...` comment for these, even though obvious.

However - since you use a lot of these constants as `u32` rather than needing them to be `NonZero*`, it could be more clear to just make them `u32` and do a static assert that they are nonzero.

Reply:
> Note that you will need a // SAFETY: ... comment for these, even though obvious.

In fact, it is not even going to compile when passing zero ... [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=fea8810d0c50836c948565a6c24453c1) ... but I'll still add some more meaningful comment.

> However - since you use a lot of these constants as u32 rather than needing them to be NonZero*, it could be
> more clear to just make them u32 and do a static assert that they are nonzero.

Hmm, eventually I'd like to convert most of these constants to module parameters to match the C implementation. Ideally, there will be an option to have module parameters of type `NonZeroXYZ` because currently accidentally panicking your kernel is as easy as `echo 0 > /sys/module/tcp_bic/parameters/max_increment`. Therefore, even though it doesn't make much sense at present, I'd leave them as `NonZeroU32` to make transitioning easier once there are the required module parameters.

Reply:
```suggestion
const BICTCP_B: NonZeroU32 = match NonZeroU32::new_unchecked(4) {
    Some(n) => n,
    None => panic!()
};
```

We can use const panic instead of unsafe.